src/github.com/getlantern/flashlight/util/http.go

```diff
@@ -2,21 +2,162 @@ package util
+	cf.dual = dual
+	cf.chained = &chainedFetcher{}
+	return cf
+}
+
+// ChainedAndFronted fetches HTTP data in parallel using both chained and fronted
+// servers.
+type chainedAndFronted struct {
+	fetcher HTTPFetcher
+	chained HTTPFetcher
+	dual    HTTPFetcher
+}
+
+// Do will attempt to execute the specified HTTP request using only a chained fetcher
+func (cf *chainedAndFronted) Do(req *http.Request) (*http.Response, error) {
+	res, err := cf.fetcher.Do(req)
+	if err != nil {
+		// If there's an error, switch back to using the dual fetcher.
+		cf.fetcher = cf.dual
+	}
+	return res, err
+}
+
+type chainedFetcher struct {
+}
+
+// Do will attempt to execute the specified HTTP request using only a chained fetcher
+func (cf *chainedFetcher) Do(req *http.Request) (*http.Response, error) {
+	log.Debugf("Using chained fronter")
+	if client, err := HTTPClient("", defaultAddr); err != nil {
+		log.Errorf("Could not create HTTP client: %v", err)
+		return nil, err
+	} else {
+		return client.Do(req)
+	}
+}
+
+type dualFetcher struct {
+	cf *chainedAndFronted
+}
+
+// Do will attempt to execute the specified HTTP request using both
+// chained and fronted servers, simply returning the first response to
+// arrive. Callers MUST use the Lantern-Fronted-URL HTTP header to
+// specify the fronted URL to use.
+func (df *dualFetcher) Do(req *http.Request) (*http.Response, error) {
+	log.Debugf("Using dual fronter")
+	frontedUrl := req.Header.Get("Lantern-Fronted-URL")
+	req.Header.Del("Lantern-Fronted-URL")
+
+	if frontedUrl == "" {
+		return nil, errors.New("Callers MUST specify the fronted URL in the Lantern-Fronted-URL header")
+	}
+	responses := make(chan *http.Response, 2)
+	errs := make(chan error, 2)
+
+	go func() {
+		client := direct.NewDirectHttpClient()
+		if r, err := http.NewRequest("GET", frontedUrl, nil); err != nil {
+			log.Errorf("Could not create request for: %v, %v", frontedUrl, err)
+			errs <- err
+		} else {
+			if resp, err := client.Do(r); err != nil {
+				log.Errorf("Could not complete request with: %v, %v", frontedUrl, err)
+				errs <- err
+			} else {
+				responses <- resp
+			}
+		}
+	}()
+	go func() {
+		if client, err := HTTPClient("", defaultAddr); err != nil {
+			log.Errorf("Could not create HTTP client: %v", err)
+			errs <- err
+		} else {
+			if resp, err := client.Do(req); err != nil {
+				log.Errorf("Could not complete request with: %v, %v", frontedUrl, err)
+				errs <- err
+			} else {
+				log.Debugf("Storing chained succeeded")
+
+				// Switch to the chained fronter since it's succeeding.
+				df.cf.fetcher = df.cf.chained
+				responses <- resp
+			}
+		}
+	}()
+
+	success := func(resp *http.Response) bool {
+		return resp.StatusCode > 199 && resp.StatusCode < 300
+	}
+
+	for i := 0; i < 2; i++ {
+		select {
+		case resp := <-responses:
+			if i == 1 {
+				log.Debugf("Got second response -- sending")
+				return resp, nil
+			} else if success(resp) {
+				log.Debugf("Got good response")
+				return resp, nil
```

Comment:
Yeah good point @fffw. We should probably shoot resp to a channel here instead and keep looping, something like that, but for now I'll just leave a comment.